### DISCUSSION TITLE
stack allocated MemoryStream

### DISCUSSION THREAD

**Discussion by Xyncgas** [dotnet/runtime#74134] [Ideas]
Although it can expand to overflow the stack, it gives 99% of object serialization an option to be allocation free simply writing the serialized result to stack allocated Span<byte>

**Comment by Xyncgas** [comment]
Maybe increase the stack size dynamically as needed, is to prevent stack overflow

**Reply by huoyaoyuan** [reply]
Stack is likely too small for `Stream` like operations.

**Reply by huoyaoyuan** [reply]
And also, dynamically adjusting stack allocated buffers would be impossible unless it's in the top frame.

**Comment by gfoidl** [comment]
> object serialization an option to be allocation free

For object serialization there are other ways than using MemoryStream, like using `IBufferWriter` (or Pipelines) so it becomes also allocation-free.

Besides that it's not to difficult to write a Stream backed by stackallocated buffer.
And for .NET's point of view I think it's similar to https://github.com/dotnet/runtime/issues/61456#issuecomment-966263731

**Reply by Clockwork-Muse** [reply]
> Besides that it's not to difficult to write a Stream backed by stackallocated buffer.

nit: you can't retain a stackallocated buffer between calls...

**Reply by gfoidl** [reply]
Why not? `ref struct`s achieve this, look at ValueStringBuilder for example.

Sure, so it's not a stream-derived type, but a steam-like type 😉.

**Reply by Clockwork-Muse** [reply]
`ValueStringBuilder` has no internal `stackalloc`.  `ref struct`s are also stack-based - the moment execution leaves the method in which they're declared they get cleaned up.

You can pass them "down" the stack as parameters to methods that get called, but the moment you try to go "up" the stack they disappear.

**Reply by gfoidl** [reply]
I don't follow what you want to express in regards of the quote.

Is your problem "backed by stackallocated buffer"? Then wrap your head around, and just pass the buffer in quite as VSB does. I didn't state anything else, or did I?
Or where's the problem here? (I hope you don't believe that I don't know how the stack is working...).

**Comment by japasc** [comment]
https://dotnet.github.io/dotNext/api/DotNext.Buffers.BufferWriterSlim-1.html

Maybe its useful here.

**Comment by Clockwork-Muse** [comment]
... in what context do you need non-allocating serialization?
